File: JavaScript/베스트앨범.js
function solution(genres, plays) {
  var answer = [];
  var chart = {}
  // 인풋값 객체화
  for (let i = 0; i < genres.length; i++) {
    if (!chart[genres[i]]) {
      chart[genres[i]] = {
        playCount: plays[i],
        index: [[plays[i], i]]
      }
    } else {
      chart[genres[i]].playCount += plays[i]
      chart[genres[i]].index.push([plays[i], i])
    }
  }
  
  // 객체 정렬
  var sorted = Object.values(chart).sort((l, r) => r.playCount - l.playCount);
  // 정렬 후 각각의 장르에서 노래들을 정렬
  sorted.forEach(genre => {
    if (genre.index.length > 1) {
      genre.index.sort((a,b) => {
        if (a[0] > b[0]) {
          return -1;
        } else if (a[0] < b[0]) {
          return 1;
        } else {
          if (a[1] > b[1]) {
            return 1
          } else {
            return -1
          }
        }
      })
      answer.push(genre.index[0][1])
      answer.push(genre.index[1][1])
    } else {
      answer.push(genre.index[0][1])
    }
  });

  return answer;
}

console.log(solution(["classic", "pop", 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500])); // [4, 1, 3, 0]

File: 프로그래머스/[level3]/베스트앨범.py
def solution(genres, plays):
    answer = []
    n = len(genres)
    album = {}
    for i in range(n):
        if album.get(genres[i]):
            album[genres[i]]['cnt'] += plays[i]
            album[genres[i]]['songs'].append((plays[i], i))
        else:
            #
            album[genres[i]] = {
                'cnt': plays[i],
                'songs': [(plays[i], i)]
            }

    # print(album)

    for key, value in album.items():
        # print(key)
        # print(value['songs'])
        album[key]['songs'] = sorted(value['songs'], key=lambda x: (-x[0], x[1]))
    ordered_album = sorted(album.items(), key=lambda x: -x[1]['cnt'])
    # print(ordered_album)
    for row in ordered_album:
        songs = row[1]['songs']
        # print(songs)
        for idx in range(len(songs)):
            if idx < 2:
                answer.append(songs[idx][1])
    # print(answer)
    return answer


print(solution(['classic', 'pop', 'classic', 'classic', 'pop'], [500, 600, 150, 800, 2500], ))  # [4, 1, 3, 0]
